src/ring_theory/I_filtration.lean

```diff
@@ -0,0 +1,181 @@
+
+-/
+
+
+universes u v
+
+variables {R M : Type u} [comm_ring R] [add_comm_group M] [module R M] (I : ideal R)
+
+open polynomial
+open_locale polynomial big_operators
+
+/-- An `I`-filtration on the module `M` is a sequence of decreasing submodules `N i` such that
+`I • N ≤ I (i + 1)`. Note that we do not require the filtration to start from `⊤`. -/
+@[ext]
+structure I_filtration (M : Type u) [add_comm_group M] [module R M] :=
+(N : ℕ → submodule R M)
+(mono : ∀ i, N (i + 1) ≤ N i)
+(smul_le : ∀ i, I • N i ≤ N (i + 1))
+
+variables (F : I_filtration I M) {I}
+
+lemma I_filtration.pow_smul_le (i j : ℕ) : I ^ i • F.N j ≤ F.N (i + j) :=
+begin
+  induction i,
+  { simp },
+  { rw [pow_succ, mul_smul, nat.succ_eq_add_one, add_assoc, add_comm 1, ← add_assoc],
+    exact (submodule.smul_mono_right i_ih).trans (F.smul_le _) }
+end
+
+lemma I_filtration.pow_smul_le_pow_smul (i j k : ℕ) : I ^ (i + k) • F.N j ≤ I ^ k • F.N (i + j) :=
+by { rw [add_comm, pow_add, mul_smul], exact submodule.smul_mono_right (F.pow_smul_le i j) }
+
+lemma I_filtration.antitone (F : I_filtration I M) : antitone F.N :=
+antitone_nat_of_succ_le F.mono
+
+/-- The trivial `I`-filtration of `N`. -/
+def trivial_I_filtration (I : ideal R) (N : submodule R M) :
+  I_filtration I M :=
+{ N := λ i, N,
+  mono := λ i, le_of_eq rfl,
+  smul_le := λ i, submodule.smul_le_right }
+
+/-- The `Sup` of a family of `I_filtration`s is an `I_filtration`. -/
+@[simps]
+def I_filtration.Sup (S : set (I_filtration I M)) : I_filtration I M :=
+{ N := Sup (I_filtration.N '' S),
+  mono := λ i, begin
+    apply Sup_le_Sup_of_forall_exists_le _,
+    rintros _ ⟨⟨_, F, hF, rfl⟩, rfl⟩,
+    exact ⟨_, ⟨⟨_, F, hF, rfl⟩, rfl⟩, F.mono i⟩,
+  end,
+  smul_le := λ i, begin
+    rw [Sup_eq_supr', supr_apply, submodule.smul_supr, supr_apply],
+    apply supr_mono _,
+    rintro ⟨_, F, hF, rfl⟩,
+    exact F.smul_le i,
+  end }
+
+/-- The `Inf` of a family of `I_filtration`s is an `I_filtration`. -/
+@[simps]
+def I_filtration.Inf (S : set (I_filtration I M)) : I_filtration I M :=
+{ N := Inf (I_filtration.N '' S),
+  mono := λ i, begin
+    apply Inf_le_Inf_of_forall_exists_le _,
+    rintros _ ⟨⟨_, F, hF, rfl⟩, rfl⟩,
+    exact ⟨_, ⟨⟨_, F, hF, rfl⟩, rfl⟩, F.mono i⟩,
+  end,
+  smul_le := λ i, begin
+    rw [Inf_eq_infi', infi_apply, infi_apply],
+    refine submodule.smul_infi_le.trans _,
+    apply infi_mono _,
+    rintro ⟨_, F, hF, rfl⟩,
+    exact F.smul_le i,
+  end }
+
+instance : partial_order (I_filtration I M) := partial_order.lift I_filtration.N I_filtration.ext
+
+instance : complete_lattice (I_filtration I M) :=
+{ sup := λ F F', ⟨F.N ⊔ F'.N, λ i, sup_le_sup (F.mono i) (F'.mono i),
+    λ i, (le_of_eq (submodule.smul_sup _ _ _)).trans $ sup_le_sup (F.smul_le i) (F'.smul_le i)⟩,
+  le_sup_left := λ a b i, le_sup_left,
+  le_sup_right := λ a b i, le_sup_right,
+  sup_le := λ a b c i j x, sup_le (i x) (j x),
+  inf := λ F F', ⟨F.N ⊓ F'.N, λ i, inf_le_inf (F.mono i) (F'.mono i),
```

Comment:
If you extract this to a separate `has_inf` then I think you might be able to golf the proof using `complete_lattice.lift` or similar.

Reply:
Maybe it's called `function.injective.complete_lattice`?